File: problems/java/greedy/max_subarray.java
class Solution {
    public int maxSubArray(int[] nums) {
        int max = nums[0];
        int curSum = nums[0];

        for(int i=1; i<nums.length; i++){
            if(curSum < 0){
                curSum = 0;
            }
            curSum += nums[i];
            max = Math.max(max, curSum);
        }

        return max;
    }
}

File: python/greedy/max_subarray.py
class Solution:
    def maxSubArray(self, nums: List[int]) -> int:
        if len(nums) == 0:
            return 0

        l, cur, res = 0, 0, float('-inf')

        for r in range(len(nums)):
            if cur < 0 and nums[r] >= cur:
                l = r
                cur = 0
            cur += nums[r]
            res = max(res, cur)

        return res